feat: collect user feedback

subzero10 opened the pull request:
## Status
**READY**

## Description
Closes: #943

## Todos
- [x] HTML Template
- [x] Form styling
- [x] Form submission (+ error control)
- [x] User Feedback Options (i.e. custom message labels)
- [x] Store and read last notice id
- [x] Add prop on react component to allow showing the feedback form automatically
- [x] Unit Tests
- [x] Integration Tests
- [x] [Documentation](https://github.com/honeybadger-io/docs/pull/240)
- [x] Upload js script to CDN
- [x] Bonus: Invalidate bunny.net (similar to how we invalidate cloudfront)

subzero10 commented on packages/core/src/client.ts:
```diff
@@ -87,6 +87,8 @@ export abstract class Client {
 
   protected abstract checkIn(id: string): Promise<void>
 
+  protected abstract showUserFeedbackForm(options: UserFeedbackForOptions): Promise<void>
```
I used the same approach we have for `checkIn`: Feature is supported only one on client side, so we throw an error if used on the server side.

subzero10 commented:
Progress!
<img width="1138" alt="image" src="https://user-images.githubusercontent.com/5813382/205434869-9cf4fe70-3743-4aa4-b474-cb26f9dc6d54.png">

subzero10 commented:
I tried not to apply custom styling so that it would be adapted based on the user's CSS.

Form:
<img width="659" alt="Screenshot 2022-12-03 at 9 41 08 PM" src="https://user-images.githubusercontent.com/5813382/205459274-f6170b74-213e-4620-b70a-86b55e514ca4.png">
Success:
<img width="627" alt="Screenshot 2022-12-03 at 9 39 49 PM" src="https://user-images.githubusercontent.com/5813382/205459286-6a9db079-da18-456b-9049-44ad4fdbd1a0.png">
Error:
<img width="659" alt="Screenshot 2022-12-03 at 9 41 08 PM" src="https://user-images.githubusercontent.com/5813382/205459295-d30b97c3-004c-4d16-bb8b-77a9bbf629af.png">

stympy commented on packages/js/assets/user-feedback-form.js:
```diff
@@ -0,0 +1,89 @@
+/**
+ * todo: this js file (bundled with the html file) should either:
```
Yup, I'm fine with creating a server-side endpoint for serving the JS

stympy replied:
Is there a reason why this JS would need to be served from API instances rather than from our JS CDN?

subzero10 replied:
Good idea! I can't think of a reason at the moment.
We could use the JS CDN. How does it work though? How can I add this as a resource/route?

subzero10 replied:
Thanks for the suggestion! I am uploading the script to our CDN! I think it's better because the js client has control over implementation + distribution.

stympy commented on packages/js/assets/user-feedback-form.js:
```diff
@@ -0,0 +1,89 @@
+    {
+      method: 'GET',
+      // todo: this should not be here
+      // server should respond with header "Access-Control-Allow-Origin: CALLER_HOST"
```
We currently only serve CORS headers on the API instances from the `/v1/notices/js` path.

subzero10 replied:
> > I tried not to apply custom styling so that it would be adapted based on the user's CSS.
> 
> I appreciate not wanting to push our own styling, but the mixture of serif (most text) and sans-serif (buttons) doesn't look great, and Times New Roman is meh. :) How about something like this (from [CSS tricks](https://css-tricks.com/snippets/css/system-font-stack/)):
> 
> ```css
> #honeybadger-feedback {
>   font-family: system-ui, "Segoe UI", Roboto, Helvetica, Arial, sans-serif, "Apple Color Emoji", "Segoe UI Emoji", "Segoe UI Symbol";
> }
> ```

Awesome, thanks!

joshuap replied:
@subzero10 I don't recall, but this feature was implemented a long time ago, so it could be a browser compatibility thing. :grin:

subzero10 replied:
OK thanks! I guess we don't have to introduce any changes since it works.
How about authentication though? Shouldn't we go for it?

joshuap replied:
> OK thanks! I guess we don't have to introduce any changes since it works.
> How about authentication though? Shouldn't we go for it?

I thought it *was* authenticated with the API key, but yes, authentication sounds reasonable. :grin:

subzero10 replied:
Nope, it's not. All you need is the notice id!

subzero10 replied:
https://github.com/honeybadger-io/honeybadger/issues/3597

stympy commented:
> I tried not to apply custom styling so that it would be adapted based on the user's CSS.

I appreciate not wanting to push our own styling, but the mixture of serif (most text) and sans-serif (buttons)  doesn't look great, and Times New Roman is meh. :) How about something like this (from [CSS tricks](https://css-tricks.com/snippets/css/system-font-stack/)):

```css
#honeybadger-feedback {
  font-family: system-ui, "Segoe UI", Roboto, Helvetica, Arial, sans-serif, "Apple Color Emoji", "Segoe UI Emoji", "Segoe UI Symbol";
}
```

subzero10 commented:
> > I tried not to apply custom styling so that it would be adapted based on the user's CSS.
> 
> I appreciate not wanting to push our own styling, but the mixture of serif (most text) and sans-serif (buttons) doesn't look great, and Times New Roman is meh. :) How about something like this (from [CSS tricks](https://css-tricks.com/snippets/css/system-font-stack/)):
> 
> ```css
> #honeybadger-feedback {
>   font-family: system-ui, "Segoe UI", Roboto, Helvetica, Arial, sans-serif, "Apple Color Emoji", "Segoe UI Emoji", "Segoe UI Symbol";
> }
> ```

Awesome, thanks!

subzero10 commented:
> 

Much, much better!
<img width="551" alt="image" src="https://user-images.githubusercontent.com/5813382/206836293-b32c7aae-1725-4949-8765-c733df8c02c6.png">

subzero10 commented:
This is finally ready! 🚀

subzero10 commented on packages/js/src/browser.ts:
```diff
@@ -40,19 +60,20 @@ class Honeybadger extends Client {
-    // Cloudflare workers do not have access to location API.
-    if (global.location != null) {
-      projectRoot = global.location.protocol + '//' + global.location.host
+  protected __afterNotifyHandlers: Types.AfterNotifyHandler[] = [
```
`afterNotify` handler to store last reported notice id, needed for the feedback form.

subzero10 commented on packages/js/assets/user-feedback-form.js:
```diff
@@ -0,0 +1,122 @@
+(function (window, document) {
+  // eslint-disable-next-line quotes
+  const TEMPLATE = `$$TEMPLATE$$`
```
The HTML template is in a separate file for better readability and maintainability. A post-build script merges the js and html files into one to be served from a script tag (see `generate-feedback-form-assets.js`).

subzero10 commented on packages/js/assets/user-feedback-form.js:
```diff
@@ -0,0 +1,122 @@
+      self.close()
+    }
+
+    const formOptions = self.getOptions()
```
The following lines allow overriding the default messages and labels.

subzero10 replied:
Good idea, separate function it is!

subzero10 commented on packages/js/assets/user-feedback-form.js:
```diff
@@ -0,0 +1,122 @@
+  HoneybadgerUserFeedbackForm.prototype.submit = function () {
+    // eslint-disable-next-line @typescript-eslint/no-this-alias
+    const self = this
+    if (self.loading) return
```
Prevent submitting more than once.

BethanyBerkowitz commented on packages/core/src/types.ts:
```diff
@@ -187,3 +187,21 @@ export type StoreContents = {
   breadcrumbs: BreadcrumbRecord[]
 }
+
+export type UserFeedbackForOptions = {
```
Should this be `UserFeedbackFormOptions`?

subzero10 replied:
Yes, thanks!

BethanyBerkowitz commented on packages/js/assets/user-feedback-form.js:
```diff
@@ -0,0 +1,122 @@
+
+    const formOptions = self.getOptions()
+    const { messages = {}, buttons = {}, labels = {} } = formOptions
+    for (let i in messages) {
```
A little confusing to use `i` here since you're looping over the object's properties and `i` usually indicates an index of an array. 

Could use something like `key` or `messageType` maybe.

BethanyBerkowitz reviewed (commented):
Cool! I had a few questions. 

Also what's the best way for me to try this out? Use the React example project?

subzero10 commented on packages/js/assets/user-feedback-form.js:
```diff
@@ -0,0 +1,122 @@
+  };
+
+  HoneybadgerUserFeedbackForm.prototype.close = function () {
+    // eslint-disable-next-line @typescript-eslint/no-this-alias
```
It's a practice I have since my early days in js scripting; maybe it's a bad practice?
The reason I still go for it is because `this` can point to a different object depending on the context (i.e. `this` inside `function () {}` vs `() => {}` (arrow function)) or even the original caller of the function.

I think I'll just disable the rule.

subzero10 commented on packages/js/test/unit/browser.test.ts:
```diff
@@ -166,4 +166,102 @@ describe('browser client', function () {
+        [JSON.stringify({ id: id2 }), { status: 201 }]
+      )
+
+      return new Promise<void>((resolve) => {
```
The `client.notify` calls are part of the returning promise (probably difficult to notice from github's interface).

subzero10 commented on packages/js/test/unit/browser.test.ts:
```diff
@@ -166,4 +166,102 @@ describe('browser client', function () {
+
+          if (notice.message === 'testing 2') {
+            expect(notice.id).toBe(id2)
+            console.log(notice.id)
```
Will remove!

subzero10 commented on packages/react/src/HoneybadgerErrorBoundary.tsx:
```diff
@@ -35,7 +41,14 @@ export default class HoneybadgerErrorBoundary extends Component<HoneybadgerError
-    this.props.honeybadger.notify(error, { context: errorInfo as never })
+    this.props.honeybadger.notify(error, {
+      context: errorInfo as never,
+      afterNotify: (error, _notice) => {
```
Good idea, will move to the constructor.

subzero10 commented:
> Cool! I had a few questions.
> 
> Also what's the best way for me to try this out? Use the React example project?

Thanks for the review, I applied all your feedback!

I also committed my changes in the example apps to try this out locally:
1. Start the express app (needed to serve the feedback form script)
  - `cd packages/js/examples/express`
  - `npm install`
  - `node index.js`
2. I used the webpack example project to test the form
  - (in a new terminal) `cd packages/js/examples/webpack`
  - `npm install`
  - uncomment line 12 in `packages/js/examples/webpack/main.js`
  - `npm run build`
  - open the `index.html` page

With a couple of tweaks, I think you should be able to use the react example app instead of the webpack one.

joshuap commented:
> > Cool! I had a few questions.
> > Also what's the best way for me to try this out? Use the React example project?
> 
> Thanks for the review, I applied all your feedback!
> 
> I also committed my changes in the example apps to try this out locally:
> 
> 1. Start the express app (needed to serve the feedback form script)
> 
> * `cd packages/js/examples/express`
> * `npm install`
> * `node index.js`
> 
> 2. I used the webpack example project to test the form
> 
> * (in a new terminal) `cd packages/js/examples/webpack`
> * `npm install`
> * uncomment line 12 in `packages/js/examples/webpack/main.js`
> * `npm run build`
> * open the `index.html` page
> 
> With a couple of tweaks, I think you should be able to use the react example app instead of the webpack one.

I tried following these instructions, and while it's setting `hb.__lastNoticeId` in the console, it's not showing the feedback form after the error is reported.

subzero10 commented:
@joshuap That's because I left out the most important part 🤦 .

You will need to call `hb.showUserFeedbackForm()` after the error is reported (after `__lastNoticeId` is set).
I updated the instructions above.

Here are a few screenshots on what you should expect in the console and network tabs of dev tools:
<img width="1601" alt="Screenshot 2023-01-05 at 9 05 45 AM" src="https://user-images.githubusercontent.com/5813382/210722668-039e83d2-e09e-495c-a871-d788f6401a0c.png">
<img width="965" alt="Screenshot 2023-01-05 at 9 06 04 AM" src="https://user-images.githubusercontent.com/5813382/210722691-67ca7067-abb4-48ba-b4ed-50cb62714ce9.png">
<img width="1605" alt="Screenshot 2023-01-05 at 9 06 30 AM" src="https://user-images.githubusercontent.com/5813382/210722714-c5dfaf11-51ae-4d3e-b224-e38db921706f.png">

BethanyBerkowitz commented:
@subzero10 I pulled your branch and did `npm run build` within the `packages/js` folder and got a few ts errors (see below). I haven't looked further into it yet, maybe you know why? 

The reason I even ran the build command is because I followed your test instructions and didn't get the error reported correctly, and I noticed that `/packages/js/examples/webpack/main.js` pulls from `/packages/js/dist`, so I figured I might need to rebuild. 

```
bethanyberkowitz@Bethanys-MBP js % npm run build

> @honeybadger-io/js@4.8.2 build
> tsc --build tsconfig.json && rollup -c && node ./scripts/copy-typedefs.js && node ./scripts/generate-feedback-form-assets.js

src/browser.ts:97:52 - error TS2694: Namespace '"/Users/bethanyberkowitz/projects/honeybadger/honeybadger-js/packages/js/node_modules/@honeybadger-io/core/build/src/types"' has no exported member 'UserFeedbackFormOptions'.

97   public async showUserFeedbackForm(options: Types.UserFeedbackFormOptions = {}) {
                                                      ~~~~~~~~~~~~~~~~~~~~~~~

test/unit/helpers.ts:30:50 - error TS2694: Namespace '"/Users/bethanyberkowitz/projects/honeybadger/honeybadger-js/packages/js/node_modules/@honeybadger-io/core/build/src/types"' has no exported member 'UserFeedbackFormOptions'.

30   protected showUserFeedbackForm(_options: Types.UserFeedbackFormOptions): Promise<void> {
                                                    ~~~~~~~~~~~~~~~~~~~~~~~


Found 2 errors.
```

subzero10 commented:
@joshuap, @BethanyBerkowitz and I just realized another thing. Probably you already know about it, but pointing it out here just in case:
- in order to test this locally, you need to serve the `index.html` file from a local web server (i.e. `localhost`, not `file://...`)
- for vs code, you can do that with a [plugin](https://marketplace.visualstudio.com/items?itemName=ritwickdey.LiveServer)
- for intellij (jetbrains), you can do that by opening the html file in the IDE and clicking on the browser icon on the top right of the file

BethanyBerkowitz commented:
Awesome, this worked for me on the webpack example project! Was able to see my feedback in HB. 

<img width="441" alt="Screen Shot 2023-01-05 at 11 49 57 AM" src="https://user-images.githubusercontent.com/16493493/210835912-d1df2395-4741-4d6d-9411-c34b3fe8a8b0.png">

joshuap commented:
It worked for me too! 👌

stympy commented:
@Pan do you want to switch to the `/v2/feedback` endpoint with authentication?

subzero10 commented:
> @subzero10 do you want to switch to the `/v2/feedback` endpoint with authentication?

I was planning to do it in a separate PR, but I tried on this one since you suggested :)

And... It's not working for some reason. Is the change deployed on the backend?
<img width="679" alt="image" src="https://user-images.githubusercontent.com/5813382/211143953-9a3ffe1f-71b2-4e2f-b39e-750cfca1b719.png">

I'm getting a 301 Redirect and then a 403. I looked at my project and verified that the comment is not added to the notice.

**Note**: I'm issuing a `GET` request (which should be OK according to [this](https://github.com/honeybadger-io/honeybadger/pull/3602/files#diff-959bc9abc46a55332bb64d5155a79323afa75a50ec1a2137ddd22d926f62c6c5R297). I can't do a `POST` request because of the implementation.

stympy commented:
Ah, it is not deployed — I added the new endpoint to the wrong repo. :) I have a PR ready, so it should be deployed soon.

stympy commented:
Ok, _now_ you should be good for testing

subzero10 commented:
Update with the latest commits:
- I switched to the v2 endpoint which requires an api key
- I made the user feedback endpoint part of the configuration options; defaults to `https://api.honeybadger.io/v2/feedback`
- I added some more error control (caught while testing the v2 endpoint)

@stympy This is ready for final review. I will merge with your approval. Thanks!

stympy reviewed (approved):
:shipit:
